### PR TITLE
Project: Linked List

### PR DESCRIPTION
**Pull request by Kaktug**
# Description

This PR contains my solution to The Odin Project's" Linked Lists project.  The details of the assignment can be viewed here:
https://www.theodinproject.com/lessons/ruby-linked-lists

> ## Assignment
> 
> You will need two classes:
> 1. `LinkedList` class, which will represent the full list.
> 1. `Node` class, containing a `#value` method and a link to the `#next_node`, set both as `nil` by default.
>
> Build the following methods in your linked list class:
> 1. `#append(value)` adds a new node containing `value` to the end of the list
> 1. `#prepend(value)` adds a new node containing `value` to the start of the list
> 1. `#size` returns the total number of nodes in the list
> 1. `#head` returns the first node in the list
> 1. `#tail` returns the last node in the list
> 1. `#at(index)` returns the node at the given `index`
> 1. `#pop` removes the last element from the list
> 1. `#contains?(value)` returns true if the passed in value is in the list and otherwise returns false.
> 1. `#find(value)` returns the index of the node containing value, or nil if not found.
> 1. `#to_s` represent your LinkedList objects as strings, so you can print them out and preview them in the console. The format should be: `( value ) -> ( value ) -> ( value ) -> nil`
>
> ## Extra Credit
> 1. `#insert_at(value, index)` that inserts a new node with the provided `value` at the given `index.`
> 1. `#remove_at(index)` that removes the node at the given `index`.
>
> **Extra Credit Tip:** When you insert or remove a node, consider how it will affect the existing nodes. Some of the nodes will need their `#next_node` link updated.

# Notes

There are far too many edge cases to worry about with this assignment so I've also spent some time reading up on [RSpec](https://rspec.info/) to implement unit tests.

One idea I had to streamline some of the code would be to create a method that hashes the nodes' index with the node itself, e.g:

```ruby
def indexed
  list = []
  node = @head

  size.times do |index|
    list[index-1] = node
    node = node.next
  end

  list
end 

linked_list = LinkedList.new
linked_list.append(10)
linked_list.append(20)
linked_list.append(30)

p linked_list.indexed
  => "0" => <Node:001 @data="10"...>, "1" => <Node:002 @data="20"...>, "2" => <Node:003 @data="30"...>
```

But, that kind of felt like cheating/outside the spirit of the exercise.

### CODE REVIEW

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      @head = @head.next
+      return
+    end
+
+    previous_node = nil
+    current_node = @head
+
+    size.times do |list_index|
+      if list_index == index
+        next_node = current_node.next.nil? ? nil : current_node.next
```
Comment (zacharydanger) [inline full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:109]: 🤔 I had to re-read this line a few times. And I think you should, too. 😅

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    current_node
+  end
+
+  def empty?
+    @head.nil?
+  end
+
+  private
+
+  def count_all_nodes(node:, counter:)
```
Comment (zacharydanger) [inline full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:146]: I'm curious why `counter` is an argument?

File: full-stack-ruby-on-rails/project-linked-list/spec/linked_list_spec.rb [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+      tail = list.tail
+      expect(tail.data).to be 20
+      expect(tail.next).to be nil
+    end
+  end
+
+  describe 'mixing data inside linked lists' do
+    it 'accepts data of any kind' do
+      list = LinkedList.new
+      var = 'i like turtles'
```
Comment (zacharydanger) [inline full-stack-ruby-on-rails/project-linked-list/spec/linked_list_spec.rb:113]: 🐢!

File: full-stack-ruby-on-rails/project-linked-list/spec/support/linked_listable.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+# A helper module to quickly create a LinkedList for use in tests
+module LinkedListable
+  def create_linked_list_with_several_appended_nodes
+    list = LinkedList.new
+    list.append(10)
+    list.append(20)
+    list.append(30)
+
+    list
```
Comment (zacharydanger) [inline full-stack-ruby-on-rails/project-linked-list/spec/support/linked_listable.rb:11]: This is an ideal candidate for [`.tap`](https://apidock.com/ruby/Object/tap)

**Review by zacharydanger — COMMENTED** [review]
Lotta good stuff in here. I like that you're writing tests!

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+              Node.new(value, @head)
+            end
+  end
+
+  def append(value)
+    node = Node.new(value)
+    empty? ? @head = node : tail.next = node
+  end
+
+  def size
```
Comment (mnishiguchi) [inline full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:26]: How about let the LinkedList object remember current size? Say default value 0, incrementing when a node is added, decrementing when a node is removed. It is just an idea in case it is useful.

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    @head = if empty?
+              Node.new(value)
+            else
+              Node.new(value, @head)
+            end
+  end
+
+  def append(value)
+    node = Node.new(value)
+    empty? ? @head = node : tail.next = node
```
Comment (mnishiguchi) [inline full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:23]: FYI: sometimes the parallel structure may help others understand what is going on quickly even though the code is short. I depends.

```rb
if empty?  
  @head = node
else
  tail.next = node
end
```

Comment (chmurph2) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:23]: I agree with Masa, his is a bit easier to scan over. I don't think I'd recommend combining ternary operators with assignment statements like that. I'd try to reserve them for simpler stuff.

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    old_tail = new_tail.next
+    new_tail.next = nil
+
+    old_tail
+  end
+
+  def contains?(value)
+    return false if empty?
+
+    !!iterate_until_data_match_or_end(node: @head, value_to_find: value)
```
Comment (chmurph2) [inline full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:59]: What do you think about renaming this `find`, and adding another method called `find?` `find?` could simply call `!!find` internally.

Comment (chmurph2) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:59]: And I see now the next method is called `find`. Nevermind, I guess? I wonder if there's room for improvement here somehow by just having the `find` vs. the need to break out `iterate_until_data_match_or_end`.

**Review by chmurph2 — APPROVED** [review]
Had some minor feedback, but overall 👍

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb (hunk 1 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:109]: Yeah, sorry Zach.  Does this seem better?
`next_node = current_node.next ||= nil`

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb (hunk 2 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:146]: I (thought) I had a reason but I've forgotten it.  I shall remove that argument and default the counter inside the `count_all_nodes` method 👍

File: full-stack-ruby-on-rails/project-linked-list/spec/support/linked_listable.rb (hunk 4 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/spec/support/linked_listable.rb:11]: That's slick, thank you!

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb (hunk 5 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:26]: Thanks Masa!  I think that is a good idea.

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb (hunk 7 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb:59]: Yeah, that makes sense.  I can just dump the `iterate_until_data_match_or_end` method and move all its logic into `find`.  Then I can do what you suggested (add a `find?` method and call that from within `contains?`)

File: full-stack-ruby-on-rails/project-linked-list/spec/linked_list_spec.rb [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    it 'adds a node to the end of the linked list' do
+      list = LinkedList.new
+      list.append(10)
+      expect(list.tail.data).to be 10
+    end
+
+    it 'adds several nodes to the end of the linked list' do
+      list = create_linked_list_with_several_appended_nodes
+      expect(list.tail.data).to be 30
+    end
```
[the remark this answers is not in the document]
Comment (zacharydanger) [reply full-stack-ruby-on-rails/project-linked-list/spec/linked_list_spec.rb]: First, I'd refactor `create_linked_list_with_several_nodes` from a method in a helper file to a `let` at the top of the spec file:

```ruby
let(:populated_list) {
  LinkedList.new.tap do |ll|
    ll.append(10)
    ll.append(20)
    ll.append(30)
  end
}

# and you can use it like this:
it "adds a new node or whatevs" do
  list = populated_list

  list.append(40)

  expect(list.tail.data).to eq 40
end
```

Though, personally, I'd just roll this up with the first test for append like:

```ruby
it "adds nodes to the end of the list" do
  list = LinkedList.new
  expect(list.tail).to be_nil

  list.append(20)
  expect(list.tail.data).to eq 20

  list.append(30)
  expect(list.tail.data).to eq 30
end
```

That way the test shows how to use `#append` and its expected behavior. Then I'd use `populated_list` elsewhere.

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb (hunk 1 above)
Comment (zacharydanger) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb]: I believe this is equivalent `next_node = current_node.next`. :D

File: full-stack-ruby-on-rails/project-linked-list/spec/linked_list_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/spec/linked_list_spec.rb]: 🙇 Thank you 🙇

File: full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb (hunk 1 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-linked-list/lib/linked_list.rb]: 😮‍💨  I know at some point this was giving me an error.  Seems like `next_node = current_node.next` works fine now, thanks Zach.
